### PR TITLE
Fix unused_result lint triggering when a function returns `()`, `!` or an empty enum

### LINKED ISSUE #43806: Unused results lint fails on trivial program
The `unused_results` lint appears to be somewhat overzealous on the current nightly compiler. I managed to reduce the issue down to a simple hello world-program, so something is definitely broken.

I tried this code:

```rust
#![deny(unused_results)]

fn main() {
    println!("hello world!");
}
```

I expected to see the code compile successfully without any warnings.

Instead, the following error message showed up.

```
error: unused result
 --> src/main.rs:4:5
  |
4 |     println!("hello world!");
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^
  |
note: lint level defined here
 --> src/main.rs:1:9
  |
1 | #![deny(unused_results)]
  |         ^^^^^^^^^^^^^^
  = note: this error originates in a macro outside of the current crate
```

## Meta

```
$ rustc --version
rustc 1.21.0-nightly (13d94d5fa 2017-08-10)
```

### PR DESCRIPTION
**Pull request by pengowen123**
Also added a test to prevent this from happening again.

Fixes #43806

### CODE REVIEW

**Comment by estebank** [comment]
r=me when the tests pass

File: src/librustc_lint/unused.rs [hunk 1]
```diff
@@ -146,6 +146,8 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnusedResults {
 
         let t = cx.tables.expr_ty(&expr);
         let ty_warned = match t.sty {
+            ty::TyTuple(ref tys, _) if tys.is_empty() => return,
+            ty::TyNever => return,
```
Comment (scottmcm) [inline src/librustc_lint/unused.rs]: Should this also apply to `enum Void {}`, and other such uninhabited types?  (Possibly like https://github.com/rust-lang/rust/pull/38069?)

Comment (pengowen123) [inline src/librustc_lint/unused.rs]: Makes sense to me. I'll add that.

**Comment by pengowen123** [comment]
I implemented @scottmcm's suggestion, so empty enums are treated the same as `!`.

**Pull request merged**